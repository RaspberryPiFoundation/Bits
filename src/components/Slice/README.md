# Slice

Not *that* kind of Slice.

A Slice is a horizontal section of the website. The will almost always contain a `.o-container` element to constrain the width of the content.

At its most basic, a Slice has some top and bottom padding, and nothing else. There are further options for altering their appearance, but the most basic implementation is as follows:

```
<div class="c-slice">
  ...
</div>
```

## Slices With Background Images

You may optionally define a `background-image`, which must be set with inline styles. Please note that **this is the only circumstance under which it is acceptable to define an inline style**.

To manage the behavior of your background image, use the `.u-bg-*` utility classes.

## Slices With Differing Background Colors

Using the background color utility classes, you can specify any background color in our palette for a Slice. You should be mindful of text contrast. Use a combination of the `.u-bg-color-*` classes and the `.c-light-on-dark` component to ensure readability.
