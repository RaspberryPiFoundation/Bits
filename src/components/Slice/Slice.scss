@import '../../styles/component-helper';

$slice-pv: $baseline-10;

.c-slice {
  background-attachment: scroll;
  background-position:   center center;
  background-repeat:     no-repeat;
  padding:               $slice-pv 0;
}

.c-slice--bg-attachment-fixed { background-attachment: fixed;        }
.c-slice--bg-attachment-local { background-attachment: local;        }
.c-slice--bg-repeat           { background-repeat:     repeat;       }
.c-slice--bg-repeat-x         { background-repeat:     repeat-x;     }
.c-slice--bg-repeat-y         { background-repeat:     repeat-y;     }
.c-slice--bg-contain          { background-size:       contain;      }
.c-slice--bg-cover            { background-size:       cover;        }
.c-slice--bg-pos-l            { background-position:   left;         }
.c-slice--bg-pos-b            { background-position:   bottom;       }
.c-slice--bg-pos-bl           { background-position:   bottom left;  }
.c-slice--bg-pos-br           { background-position:   bottom right; }
.c-slice--bg-pos-r            { background-position:   right;        }
.c-slice--bg-pos-t            { background-position:   top;          }
.c-slice--bg-pos-tl           { background-position:   top left;     }
.c-slice--bg-pos-tr           { background-position:   top right;    }
