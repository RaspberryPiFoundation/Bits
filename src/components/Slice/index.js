import { Base } from 'react-iotacss'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

import './Slice.scss'

export const Slice = ({ children, className, ...props }) => {
  const classNames = classnames('c-slice', className)

  return (
    <Base className={classNames} {...props}>
      {children}
    </Base>
  )
}

Slice.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
}

export default Slice
