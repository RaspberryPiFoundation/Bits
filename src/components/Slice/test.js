import { shallow } from 'enzyme'
import Slice from '.'
import React from 'react'

let wrapper

const children = <span>Whatever</span>

describe('<Slice />', () => {
  beforeEach(() => {
    wrapper = shallow(<Slice>{children}</Slice>)
  })

  describe('Renders', () => {
    it('without crashing', () => {
      expect(wrapper.find('.c-slice')).toBePresent()
    })

    it('matches its snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Variants', () => {
    describe('...white', () => {
      beforeEach(() => {
        wrapper = shallow(<Slice white>{children}</Slice>)
      })

      it('links to the right place', () => {
        expect(wrapper.find('.c-slice')).toHaveClassName('.c-slice--white')
      })
    })
  })
})
