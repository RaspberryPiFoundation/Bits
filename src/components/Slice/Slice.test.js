import { shallow } from 'enzyme'
import Slice from '../Slice'
import React from 'react'

let slice

const children = 'Whatever'

describe('<Slice />', () => {
  describe('Renders...', () => {
    beforeEach(() => {
      slice = shallow(<Slice>{children}</Slice>)
    })

    it('without crashing', () => {
      expect(slice.find('.c-slice')).toExist()
    })

    it('matches its snapshot', () => {
      expect(slice).toMatchSnapshot()
    })
  })

  describe('Modifiers', () => {
    describe('white', () => {
      it('adds the correct className', () => {
        slice = shallow(<Slice white>{children}</Slice>)
        expect(slice.find('.c-slice')).toHaveClassName('c-slice--white')
      })
    })
  })
})
