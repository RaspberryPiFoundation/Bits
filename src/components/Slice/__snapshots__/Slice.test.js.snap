// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Player /> Renders... matches its snapshot 1`] = `
<Base
  className="c-slice"
  tagName="div"
>
  Whatever
</Base>
`;
